File: src/main/java/org/wysko/midis2jam2/gui/LauncherSettings.java
/*
 * Copyright (C) 2021 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.wysko.midis2jam2.gui;

import org.wysko.midis2jam2.M2J2Settings.InstrumentTransition;

import javax.swing.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LauncherSettings {
	
	private String lastMidiDir;
	private List<String> soundFontPaths;
	private InstrumentTransition transition;
	private String midiDevice;
	private boolean fullscreen;
	private final Map<String, Integer> deviceLatencyMap;
	private boolean legacyDisplay;
	
	public int getLatencyForDevice(String deviceName) {
		var integer = deviceLatencyMap.get(deviceName);
		return integer == null ? 0 : integer;
	}
	
	public void setLatencyForDevice(String deviceName, int value) {
		deviceLatencyMap.put(deviceName, value);
	}
	
	public boolean isFullscreen() {
		return fullscreen;
	}
	
	public void setFullscreen(boolean fullscreen) {
		this.fullscreen = fullscreen;
	}
	
	public LauncherSettings() {
		lastMidiDir = new JFileChooser().getFileSystemView().getDefaultDirectory().getAbsolutePath();
		soundFontPaths = new ArrayList<>();
		transition = InstrumentTransition.NORMAL;
		midiDevice = "Gervill";
		fullscreen = false;
		deviceLatencyMap = new HashMap<>();
		deviceLatencyMap.put("Gervill", 100);
		legacyDisplay = false;
	}
	
	public String getLastMidiDir() {
		return lastMidiDir;
	}
	
	public void setLastMidiDir(String lastMidiDir) {
		this.lastMidiDir = lastMidiDir;
	}
	
	public List<String> getSoundFontPaths() {
		return soundFontPaths;
	}
	
	public void setSoundFontPaths(List<String> soundFontPaths) {
		this.soundFontPaths = soundFontPaths;
	}
	
	public InstrumentTransition getTransition() {
		return transition;
	}
	
	public void setTransition(InstrumentTransition transition) {
		this.transition = transition;
	}
	
	public String getMidiDevice() {
		return midiDevice;
	}
	
	public void setMidiDevice(String midiDevice) {
		this.midiDevice = midiDevice;
	}
	
	@Override
	public String toString() {
		return "LauncherSettings{" +
				"lastMidiDir='" + lastMidiDir + '\'' +
				", soundFontPaths=" + soundFontPaths +
				", transition=" + transition +
				", midiDevice='" + midiDevice + '\'' +
				'}';
	}
	
	public boolean isLegacyDisplay() {
		return legacyDisplay;
	}
	
	public void setLegacyDisplay(boolean legacyDisplay) {
		this.legacyDisplay = legacyDisplay;
	}
}

File: src/main/java/org/wysko/midis2jam2/gui/LauncherSettings.kt
/*
 * Copyright (C) 2021 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.wysko.midis2jam2.gui

import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File

/**
 * Defines settings that are configurable in the launcher.
 */
@Serializable
data class LauncherSettings(
    /** The last directory that the user selected a MIDI file from. Defaults to the user's home directory. */
    var lastMidiDir: String = System.getProperty("user.home"),

    /** A list of paths to SoundFonts. */
    var soundFontPaths: MutableList<String> = mutableListOf(),

    /** The currently selected MIDI device. */
    var midiDevice: String = "Gervill",

    /** True if midis2jam2 should run in fullscreen mode, false otherwise. */
    var fullscreen: Boolean = false,

    /** Stores the amount of latency fix each MIDI device should have. */
    var deviceLatencyMap: MutableMap<String, Int> = mutableMapOf("Gervill" to 0),

    /** True if the autocam should be enabled when the song starts, false otherwise. */
    var autoAutoCam: Boolean = true,

    /** True if the legacy display engine should be used, false otherwise. */
    var isLegacyDisplay: Boolean = false,

    /** The locale of the launcher. */
    var locale: String = "en"
) {
    /** Writes the settings to disk. */
    fun save() {
        launcherSettingsFile().writeText(Json.encodeToString(this))
    }
}

/** Returns the file that contains the settings for the launcher. */
private fun launcherSettingsFile(): File = File(
    "${System.getProperty("user.home")}${System.getProperty("file.separator")}midis2jam2.settings"
)

/**
 * Loads the launcher settings, stored at [launcherSettingsFile]. When loading, it also cleans out any unavailable
 * SoundFont files.
 */
internal fun loadLauncherSettingsFromFile(): LauncherSettings {
    return if (launcherSettingsFile().exists()) {
        try {
            Json.decodeFromString<LauncherSettings>(launcherSettingsFile().readText()).also { settings ->
                settings.soundFontPaths = settings.soundFontPaths.filter { File(it).exists() }.toMutableList()
            }
        } catch (e: Exception) {
            LauncherSettings()
        }
    } else {
        LauncherSettings()
    }
}